granulate_utils/futures.py

```diff
@@ -0,0 +1,38 @@
+from concurrent.futures import Future, ThreadPoolExecutor, _base, as_completed
+from contextlib import contextmanager
+from typing import Callable, Collection, Generator, TypeVar
+
+T = TypeVar("T")
+
+
+@contextmanager
+def wrap_thread_pool(pool: ThreadPoolExecutor) -> Generator[ThreadPoolExecutor, None, None]:
+    try:
+        yield pool
+    except _base.TimeoutError as e:
+        # Translate _base.TimeoutError to be a standard TimeoutError, this was only fixed in python 3.11
+        # so we achieve here the same behavior for older python versions
+        # https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.TimeoutError
+        raise TimeoutError(*e.args) from e
+    finally:
+        pool.shutdown(wait=False)
+
+
+def call_in_parallel(
+    callables: Collection[Callable[[], T]], timeout: float, max_threads: int = 10
+) -> Generator[Future[T], None, None]:
```

Comment:
I tried to run it with gProfiler - 
```
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/app/gprofiler/__main__.py", line 5, in <module>
    from gprofiler.main import main
  File "/app/gprofiler/main.py", line 46, in <module>
    from gprofiler.metadata.metadata_collector import get_current_metadata, get_static_metadata
  File "/app/gprofiler/metadata/metadata_collector.py", line 4, in <module>
    from granulate_utils.metadata.cloud import get_static_cloud_instance_metadata
  File "/usr/local/lib/python3.8/dist-packages/granulate_utils/metadata/cloud.py", line 16, in <module>
    from granulate_utils.futures import call_in_parallel
  File "/usr/local/lib/python3.8/dist-packages/granulate_utils/futures.py", line 23, in <module>
    ) -> Generator[Future[T], None, None]:
TypeError: 'type' object is not subscriptable
```

it's 3.8 (gProfiler container)

Reply:
Ahh shit, we should run the tests on all python versions 😢

Reply:
We do run linting on python 3.8 but apparently this isn't caught

Reply:
Using version matrix for testing as well now so we don't accidentally introduce bugs missed by the linter